lib/ansible/plugins/inventory/foreman.py

```diff
@@ -113,77 +120,29 @@ def _get_session(self):
-                self._cache[self.cache_key] = {'url': ''}
-
-            results = []
-            s = self._get_session()
-            params = {'page': 1, 'per_page': 250}
-            while True:
-                ret = s.get(url, params=params)
-                if ignore_errors and ret.status_code in ignore_errors:
-                    break
-                ret.raise_for_status()
-                json = ret.json()
-
-                # process results
-                # FIXME: This assumes 'return type' matches a specific query,
-                #        it will break if we expand the queries and they dont have different types
-                if 'results' not in json:
-                    # /hosts/:id dos not have a 'results' key
-                    results = json
-                    break
-                elif isinstance(json['results'], MutableMapping):
-                    # /facts are returned as dict in 'results'
-                    results = json['results']
-                    break
-                else:
-                    # /hosts 's 'results' is a list of all hosts, returned is paginated
-                    results = results + json['results']
-
-                    # check for end of paging
-                    if len(results) >= json['subtotal']:
-                        break
-                    if len(json['results']) == 0:
-                        self.display.warning("Did not make any progress during loop. expected %d got %d" % (json['subtotal'], len(results)))
-                        break
-
-                    # get next page
-                    params['page'] += 1
-
-            self._cache[self.cache_key][url] = results
-
-        return self._cache[self.cache_key][url]
-
-    def _get_hosts(self):
-        return self._get_json("%s/api/v2/hosts" % self.foreman_url)
-
-    def _get_all_params_by_id(self, hid):
-        url = "%s/api/v2/hosts/%s" % (self.foreman_url, hid)
-        ret = self._get_json(url, [404])
-        if not ret or not isinstance(ret, MutableMapping) or not ret.get('all_parameters', False):
-            ret = []
-        return ret.get('all_parameters')
-
-    def _get_facts_by_id(self, hid):
-        url = "%s/api/v2/hosts/%s/facts" % (self.foreman_url, hid)
-        return self._get_json(url)
-
-    def _get_facts(self, host):
-        """Fetch all host facts of the host"""
-
-        ret = self._get_facts_by_id(host['id'])
-        if len(ret.values()) == 0:
-            facts = {}
-        elif len(ret.values()) == 1:
-            facts = list(ret.values())[0]
-        else:
-            raise ValueError("More than one set of facts returned for '%s'" % host)
-        return facts
+    def _get_all_hosts(self):
+
+        s = self._get_session()
+
+        if self.cache:
+            inventory = self.cache.get_inventory()
+
+            if inventory is not None:
+                return inventory
+
+        params = {'page': 1, 'per_page': 1000}
```

Comment:
This limits you to 1000 hosts where previously it iterated all the pages.

Reply:
Yeah, this was a dirty workaround because we had a big showstopper here. It's on my ToDo to clean up the current state.